### ISSUE TITLE
Exclusive Canonicalization transform "http://www.w3.org/TR/2001/REC-xml-c14n-20010315#WithComments" algorithm is not supported by ADFS 4.0

### ISSUE THREAD

**Issue by innocentliny** [onelogin/java-saml#145]
Hi,

I setup a SP and using servlet to generate metedata xml by referencing java-saml/samples/java-saml-tookit-jspsample/src/main/webapp/metadata.jsp.
I saved the metadata as xml file and import to ADFS 4.0 by adding relaying party trust function, then the error is as below:
![image](https://user-images.githubusercontent.com/5312376/35953796-4d854af8-0cc1-11e8-87e1-0568f76cd8f8.png)
It says "Exclusive Canonicalization transform "http://www.w3.org/TR/2001/REC-xml-c14n-20010315#WithComments" algorithm is not supported".

I checked the [http://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf](http://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf) at line 2941:

> Signatures in SAML messages SHOULD NOT contain transforms other than the enveloped signature
transform (with the identifier http://www.w3.org/2000/09/xmldsig#enveloped-signature) or the exclusive canonicalization transforms (with the identifier http://www.w3.org/2001/10/xml-exc-c14n# or
http://www.w3.org/2001/10/xml-exc-c14n#WithComments).

If I am right, "http://www.w3.org/TR/2001/REC-xml-c14n-20010315#WithComments" is added in com.onelogin.saml2.util.Util line 1087:
`transforms.addTransform(c14nMethod);`

Is it a correct code? It seems ADFS 4.0 doesn't accept it.
Or I need to config ADFS to accept it?

Please let me know if any comments, thank you very much.

**Comment by dave-handy** [comment]
I just ran into this as well.. does this mean that java-saml just doesn't support ADFS?

**Comment by pitbulk** [comment]
Replace
```
String c14nMethod = Constants.C14N_WC;
```
by
```
String c14nMethod = Constants.C14NEXC_WC;
```

And let me know if that fixes your issue

**Comment by dave-handy** [comment]
I deployed this and now we just have "ID6013: The signature verification failed." from ADFS. I am not sure if that means we solved the issue, and now it's running into a separate issue, or it's not canonicalizing properly according to this new method (the library code seems like it should work just fine).

I ran my generated metadata files through https://www.aleksey.com/xmlsec/xmldsig-verifier.html and it failed both before and after this change, so I'm a bit confused. I typically integrate with okta and onelogin and they don't support specifying a metadata file or URL so I don't know how else to check these are being signed properly.

Let me know if you have other ideas! It seems like I will probably need to turn off metadata signing.

**Comment by pitbulk** [comment]
Let me spend some time tomorrow to review and I will let you know what I discover.

**Comment by dave-handy** [comment]
I believe we found the issue with the signature verification, in that the cert my team is using is self signed, and it seems that ADFS performs a revocation and trust chain verification on the signer cert by default, unlike other IDPs we've worked with. I am still using the patched version of java-saml with your suggested change to the `c14nMethod` value, which does appear to be necessary to even get this far.

TLDR; I think the change is working, but you probably want some additional verification ;)
I pushed up a PR with the change
